addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/discovery/HomematicDeviceDiscoveryService.java

```diff
@@ -59,53 +65,138 @@ public void deactivate() {
+        }
+    }
+    
+    private void setIsInInstallMode() {
+        synchronized (installModeSync) {
+            isInInstallMode = true;
+        }
+    }
+    
+    private void waitForInstallmodeFinished(int timeout) throws InterruptedException {
+        synchronized (installModeSync) {
+            while (isInInstallMode) {
+                installModeSync.wait(timeout);
+            }
+        }        
+    }
+    
+    private void waitForLoadDevicesFinished() throws InterruptedException, ExecutionException {
+        if (loadDevicesFuture != null) {
+            loadDevicesFuture.get();
+        }
+    }
 
     /**
      * Waits for the discovery scan to finish and then returns.
      */
     public void waitForScanFinishing() {
-        if (scanFuture != null) {
-            logger.debug("Waiting for finishing Homematic device discovery scan");
-            try {
-                scanFuture.get();
-                logger.debug("Homematic device discovery scan finished");
-            } catch (CancellationException ex) {
-                // ignore
-            } catch (Exception ex) {
-                logger.error("Error waiting for device discovery scan: {}", ex.getMessage(), ex);
-            }
-        }
+    
+        logger.debug("Waiting for finishing Homematic device discovery scan"); 
+        try {
+            waitForInstallmodeFinished(DISCOVER_TIMEOUT_SECONDS * 1000);
+            waitForLoadDevicesFinished();    
+                        
+        } catch (ExecutionException | InterruptedException ex) {
+            // ignore
+        } catch (Exception ex) {
+            logger.error("Error waiting for device discovery scan: {}", ex.getMessage(), ex);
+        }        
+        
+        String gatewayId = bridgeHandler != null ? bridgeHandler.getGateway().getId() : "UNKNOWN";
+        logger.debug("Finished Homematic device discovery scan on gateway '{}'", gatewayId);
     }
 
     /**
      * Starts a thread which loads all Homematic devices connected to the gateway.
      */
     public void loadDevices() {
-        if (scanFuture == null && bridgeHandler.getGateway() != null) {
-            scanFuture = scheduler.submit(() -> {
-                try {
-                    final HomematicGateway gateway = bridgeHandler.getGateway();
-                    gateway.loadAllDeviceMetadata();
-                    bridgeHandler.getTypeGenerator().validateFirmwares();
-                    logger.debug("Finished Homematic device discovery scan on gateway '{}'", gateway.getId());
-                } catch (Throwable ex) {
-                    logger.error("{}", ex.getMessage(), ex);
-                } finally {
-                    scanFuture = null;
-                    bridgeHandler.setOfflineStatus();
-                    removeOlderResults(getTimestampOfLastScan());
+        if (loadDevicesFuture == null && bridgeHandler.getGateway() != null) {
+            loadDevicesFuture = scheduler.submit(new Runnable() {
```

Comment:
Is there a particular reason for using an anonymous class instead of a lambda expression here? I like the original version better since it requires less boilerplate code.

Reply:
Mhh, actually there is no specific reason for this. Propably I was playing around with sth else here and didn't rollback properly. Therefore I will return to lambda expression here.